optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/decorator/WeightFactorySelectionSorter.java

```diff
@@ -34,13 +34,15 @@
  * Sorts a selection {@link List} based on a {@link SelectionSorterWeightFactory}.
  * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
  * @param <T> the selection type
+ * @param <W> the weight type
  */
-public class WeightFactorySelectionSorter<Solution_, T> implements SelectionSorter<Solution_, T> {
+public class WeightFactorySelectionSorter<Solution_, T, W extends Comparable<W>>
+        implements SelectionSorter<Solution_, T> {
 
-    private final SelectionSorterWeightFactory<Solution_, T> selectionSorterWeightFactory;
-    private final Comparator<Comparable> appliedWeightComparator;
+    private final SelectionSorterWeightFactory<Solution_, T, W> selectionSorterWeightFactory;
+    private final Comparator<W> appliedWeightComparator;
```

Comment:
The impact on `WeightFactorySelectionSorter` is a nice side effect of the additional 3rd parameter.

The `WeightFactory` itself is complex because it's an abstract thing. If we add the 3rd parameter, the complexity remains the same but it becomes more visible and easier to understand, which is a good thing.

The downside is that it will slightly hurt users with existing weight factory implementations. They will need to add the weight type to the `implements` declaration and may need to change the return type of `createSorterWeight()`. On the other hand it will help users that are going to implement a new weight factory.

So in my opinion this is an improvement and it's better to make small improvements even at some cost (when the time is right).